cpp/src/dual_simplex/mip_node.hpp

```diff
@@ -230,11 +242,102 @@ void remove_fathomed_nodes(std::vector<mip_node_t<i_t, f_t>*>& stack)
 template <typename i_t, typename f_t>
 class node_compare_t {
  public:
-  bool operator()(mip_node_t<i_t, f_t>& a, mip_node_t<i_t, f_t>& b)
+  bool operator()(mip_node_t<i_t, f_t>& a, mip_node_t<i_t, f_t>& b) const
   {
     return a.lower_bound >
            b.lower_bound;  // True if a comes before b, elements that come before are output last
   }
+
+  bool operator()(mip_node_t<i_t, f_t>* a, mip_node_t<i_t, f_t>* b) const
+  {
+    return a->lower_bound >
+           b->lower_bound;  // True if a comes before b, elements that come before are output last
+  }
+};
+
+template <typename i_t, typename f_t>
+class search_tree_t {
+ public:
+  search_tree_t(f_t root_lower_bound, const std::vector<variable_status_t>& basis, logger_t& log)
+    : root(root_lower_bound, basis), num_nodes(0), log(log)
+  {
+  }
+
+  search_tree_t(mip_node_t<i_t, f_t>&& node, logger_t& log)
+    : root(std::forward<mip_node_t<i_t, f_t>&&>(node)), num_nodes(0), log(log)
+  {
+  }
+
+  f_t get_lower_bound() const { return root.lower_bound; }
+
+  void update_tree(mip_node_t<i_t, f_t>* node_ptr, node_status_t status)
+  {
+    std::lock_guard<std::mutex> lock(mutex);
+    std::vector<mip_node_t<i_t, f_t>*> stack;
+    node_ptr->set_status(status, stack);
+    remove_fathomed_nodes(stack);
+  }
+
+  void branch(mip_node_t<i_t, f_t>* parent_node,
+              i_t branch_var,
+              f_t branch_var_val,
+              const std::vector<variable_status_t>& parent_vstatus,
+              const lp_problem_t<i_t, f_t>& original_lp)
+  {
+    i_t id;
+
+#pragma omp atomic capture
```

Comment:
We are using a mix of std::thread and openmp.   Can we move to using only `std::thread` ?